src/memray/_memray.pyx

```diff
@@ -318,97 +321,130 @@ def start_thread_trace(frame, event, arg):
-        if not self._get_reader().isOpen():
+    cdef void _ensure_not_closed(self) except *:
+        if self._file is None:
             raise ValueError("Operation on a closed FileReader")
 
     @property
     def closed(self):
-        return not self._get_reader().isOpen()
+        return self._file is None
 
     def __enter__(self):
         return self
 
     def __exit__(self, exc_type, exc_value, exc_traceback):
         self.close()
 
-    def __dealloc__(self):
-        self._reader.reset()
-    
-    def _yield_allocations(self, size_t index, merge_threads):
-        for elem in Py_GetSnapshotAllocationRecords(
-            self._allocations, index, merge_threads):
+    def _yield_allocations(self, size_t max_records, bool merge_threads):
+        cdef SnapshotAllocationAggregator aggregator
+        cdef shared_ptr[RecordReader] reader_sp = make_shared[RecordReader](
+            unique_ptr[FileSource](new FileSource(self._path))
+        )
+        cdef RecordReader* reader = reader_sp.get()
+
+        while max_records > 0:
+            PyErr_CheckSignals()
+            ret = reader.nextRecord()
+            if ret == RecordResult.RecordResultAllocationRecord:
+                aggregator.addAllocation(reader.getLatestAllocation())
+                max_records -= 1
+            elif ret == RecordResult.RecordResultMemoryRecord:
+                pass
+            else:
+                break
+
+        for elem in Py_ListFromSnapshotAllocationRecords(
+            aggregator.getSnapshotAllocations(merge_threads)
+        ):
             alloc = AllocationRecord(elem)
-            (<AllocationRecord> alloc)._reader = self._reader
+            (<AllocationRecord> alloc)._reader = reader_sp
             yield alloc
-            self._ensure_reader_is_open()
 
-    cdef inline HighWatermark* _get_high_watermark(self) except*:
-        cdef HighWatermarkFinder finder
-        if self._high_watermark == NULL:
-            for record in self._allocations:
-                finder.processAllocation(record)
-            self._high_watermark = make_unique[HighWatermark](finder.getHighWatermark())
-        return self._high_watermark.get()
+        reader.close()
 
     def get_high_watermark_allocation_records(self, merge_threads=True):
-        self._ensure_reader_is_open()
-        cdef HighWatermark* watermark = self._get_high_watermark()
-        yield from self._yield_allocations(watermark.index, merge_threads)
+        self._ensure_not_closed()
+        yield from self._yield_allocations(
+            self._high_watermark.index + 1,
+            merge_threads,
+        )
 
     def get_leaked_allocation_records(self, merge_threads=True):
-        self._ensure_reader_is_open()
-        cdef size_t snapshot_index = self._allocations.size() - 1
-        yield from self._yield_allocations(snapshot_index, merge_threads)
+        self._ensure_not_closed()
+        yield from self._yield_allocations(<size_t>(-1), merge_threads)
```

Comment:
Please, add a comment for those that are not part of the selected cabal of "we know how unsigned overflow works" as we are :)

Reply:
Would you prefer `numeric_limits[size_t].max()`? It's an extra `cimport` but might be more readable.

Reply:
On reflection, I prefer that. I'll go with that.